### ISSUE TITLE
Create API definition from provided dependency in ear

### ISSUE THREAD

**Issue by saschakr** [smallrye/smallrye-open-api#785]
When using a provided dependency in war, the OpenAPI definition is not complete.

Example: 
- example-war 
  - provided dependency on example-jar 
- example-jar
- example-ear
  - bundles both

Expected outcome for /openapi would contains both example schemas

```yaml
components:
  schemas:
    ExampleFromWAR:
      type: object
      properties:
        name:
          type: string
    ExampleFromJAR:
      type: object
      properties:
        name:
          type: string
```

but only the class from the war will be displayed.

A dummy project can be found here:  https://github.com/saschakr/open-api-example
Is there any posibility to fix this?

**Comment by MikeEdgar** [comment]
Hi @saschakr - are you using Wildfly?

**Comment by saschakr** [comment]
@MikeEdgar Yes, I'm using Wildfly 22.0.1Final.

**Comment by MikeEdgar** [comment]
@pferraro - I think this might need to be addressed in Wildfly. What do you think?

**Comment by pferraro** [comment]
@saschakr  Modules of an EAR are not implicitly visible to each other.  Does you example-war include example-jar in its classpath?
See Section 8.2 of the Jakarta EE platform specification for details.  If configured correctly, this should already work.

**Comment by MikeEdgar** [comment]
@saschakr - were you able to get this working ?

**Comment by saschakr** [comment]
When building the the war without "provided", the example-jar will show up in the Class-Path Section of the MANIFEST.MF.
Adding "provided" will remove it.

Is there any way to fix this without removing the provided in the war's pom.xml?

**Comment by MikeEdgar** [comment]
You should be able to use the maven-jar-plugin to add Class-Path entries to the MANIFEST.MF. See if this helps: https://maven.apache.org/shared/maven-archiver/examples/classpath.html#Add

**Comment by saschakr** [comment]
Im using this from maven-war-plugin (see [L53](https://github.com/saschakr/open-api-example/blob/main/example-war/pom.xml#L53) and the jar is still packaged.

**Comment by MikeEdgar** [comment]
See if this works:
```xml
    <dependency>
      <groupId>org.example</groupId>
      <artifactId>example-jar</artifactId>
      <version>1.0-SNAPSHOT</version>
      <scope>compile</scope>
      <optional>true</optional>
    </dependency>
```
And:
```xml
      <plugin>
        <artifactId>maven-war-plugin</artifactId>
        <version>2.1.1</version>
        <configuration>
          <archive>
            <manifest>
              <addClasspath>true</addClasspath>
              <classpathPrefix>lib/</classpathPrefix>
            </manifest>
          </archive>
        </configuration>
      </plugin>
```

**Comment by MikeEdgar** [comment]
@saschakr - any luck using the configuration I mentioned?

**Comment by grimmjo** [comment]
I've run into the same problem. The JakartaEE 8 specification(https://jakarta.ee/specifications/platform/8/platform-spec-8.html#a2948) says:

> All files in this directory (but not subdirectories) with a .jar extension must be made available to all components packaged in the EAR file, including application clients. These libraries may reference other libraries, either bundled with the application or installed separately, using any of the techniques described herein.

The application in my case is working correctly, just the OpenAPI definition is missing for the classes from the ear's lib directory and the other modules.

I also run into the following issue, having a "missing" class annotated with @BeanParam.

```
2021-06-15 09:28:53,093 ERROR [                                       org.jboss.msc.service.fail] (MSC service thread 1-7 ) [u: j: s: h:] MSC000001: Failed to start service org.wildfly.undertow.host.default-server.default-host./api/v1/offlinemedia/openapi: org.jboss.msc.service.StartException in service org.wildfly.undertow.host.default-server.default-host./api/v1/openapi: java.lang.NullPointerException
	at org.wildfly.clustering.service@24.0.0.Beta1//org.wildfly.clustering.service.FunctionalService.start(FunctionalService.java:66)
	at org.jboss.msc@1.4.12.Final//org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1739)
	at org.jboss.msc@1.4.12.Final//org.jboss.msc.service.ServiceControllerImpl$StartTask.execute(ServiceControllerImpl.java:1701)
	at org.jboss.msc@1.4.12.Final//org.jboss.msc.service.ServiceControllerImpl$ControllerTask.run(ServiceControllerImpl.java:1559)
	at org.jboss.threads@2.4.0.Final//org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
	at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1363)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at io.smallrye.openapi//io.smallrye.openapi.runtime.util.JandexUtil.inheritanceChain(JandexUtil.java:623)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.dataobject.TypeResolver.forClass(TypeResolver.java:426)
	at io.smallrye.openapi//io.smallrye.openapi.jaxrs.JaxRsParameterProcessor.readJaxRsParameter(JaxRsParameterProcessor.java:214)
	at io.smallrye.openapi//io.smallrye.openapi.jaxrs.JaxRsParameterProcessor.readAnnotatedType(JaxRsParameterProcessor.java:170)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.spi.AbstractParameterProcessor.readAnnotatedType(AbstractParameterProcessor.java:877)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.spi.AbstractParameterProcessor.processOperationParameters(AbstractParameterProcessor.java:259)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.spi.AbstractParameterProcessor.process(AbstractParameterProcessor.java:243)
	at io.smallrye.openapi//io.smallrye.openapi.jaxrs.JaxRsParameterProcessor.process(JaxRsParameterProcessor.java:82)
	at io.smallrye.openapi//io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethod(JaxRsAnnotationScanner.java:421)
	at io.smallrye.openapi//io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.lambda$processResourceMethods$0(JaxRsAnnotationScanner.java:271)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at io.smallrye.openapi//io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethods(JaxRsAnnotationScanner.java:269)
	at io.smallrye.openapi//io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClass(JaxRsAnnotationScanner.java:239)
	at io.smallrye.openapi//io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClasses(JaxRsAnnotationScanner.java:214)
	at io.smallrye.openapi//io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.scan(JaxRsAnnotationScanner.java:146)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:127)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.OpenApiProcessor.modelFromAnnotations(OpenApiProcessor.java:136)
	at org.wildfly.extension.microprofile.openapi-smallrye@24.0.0.Beta1//org.wildfly.extension.microprofile.openapi.deployment.OpenAPIModelServiceConfigurator.get(OpenAPIModelServiceConfigurator.java:175)
	at org.wildfly.extension.microprofile.openapi-smallrye@24.0.0.Beta1//org.wildfly.extension.microprofile.openapi.deployment.OpenAPIModelServiceConfigurator.get(OpenAPIModelServiceConfigurator.java:96)
	at org.wildfly.clustering.service@24.0.0.Beta1//org.wildfly.clustering.service.FunctionalService.start(FunctionalService.java:63)
	... 8 more
```

Maybe the stacktrace helps you tackling the problem :)

**Comment by MikeEdgar** [comment]
@grimmjo - is the type that is annotated with `@BeanParam` located in a jar in the EAR's `lib/`?

@pferraro, it seems that the class-path is setup fine in the examples (based on the comments), but somehow the Jandex index does not contain the class information.

**Comment by grimmjo** [comment]
@MikeEdgar yes the Type is in the EAR's lib

**Comment by pferraro** [comment]
@MikeEdgar I'll file an issue against WildFly.

**Comment by grimmjo** [comment]
@pferraro 
Any news on the issue against wildfly? Can you provide an issue ID or link?

**Comment by grimmjo** [comment]
So the Nullpointer is fixed with #835. 

@pferraro and @MikeEdgar 
I've a solution for the class loading issue. I'll open an issue in wildfly's Jira and link it here. I'll also provide a PR in the wildfly repo and also link it here.

**Comment by MikeEdgar** [comment]
@saschakr , @grimmjo - closing this issue since the fix has been applied in WildFly.
